libsolidity/formal/CHC.cpp

```diff
@@ -43,53 +43,31 @@ using namespace solidity::frontend;
 CHC::CHC(
 	smt::EncodingContext& _context,
 	ErrorReporter& _errorReporter,
-	map<util::h256, string> const& _smtlib2Responses,
-	ReadCallback::Callback const& _smtCallback,
-	[[maybe_unused]] smtutil::SMTSolverChoice _enabledSolvers
+	[[maybe_unused]] map<util::h256, string> const& _smtlib2Responses,
+	[[maybe_unused]] ReadCallback::Callback const& _smtCallback,
+	smtutil::SMTSolverChoice _enabledSolvers
 ):
 	SMTEncoder(_context),
 	m_outerErrorReporter(_errorReporter),
 	m_enabledSolvers(_enabledSolvers)
 {
-#ifdef HAVE_Z3
-	if (_enabledSolvers.z3)
-		m_interface = make_unique<smtutil::Z3CHCInterface>();
+	bool usesZ3 = _enabledSolvers.z3;
+#ifndef HAVE_Z3
+	usesZ3 = false;
 #endif
-	if (!m_interface)
+	if (!usesZ3)
 		m_interface = make_unique<smtutil::CHCSmtLib2Interface>(_smtlib2Responses, _smtCallback);
 }
 
 void CHC::analyze(SourceUnit const& _source)
 {
 	solAssert(_source.annotation().experimentalFeatures.count(ExperimentalFeature::SMTChecker), "");
 
-	bool usesZ3 = false;
-#ifdef HAVE_Z3
-	usesZ3 = m_enabledSolvers.z3;
-	if (usesZ3)
-	{
-		auto z3Interface = dynamic_cast<smtutil::Z3CHCInterface const*>(m_interface.get());
-		solAssert(z3Interface, "");
-		m_context.setSolver(z3Interface->z3Interface());
-	}
-#endif
-	if (!usesZ3)
-	{
-		auto smtlib2Interface = dynamic_cast<smtutil::CHCSmtLib2Interface const*>(m_interface.get());
-		solAssert(smtlib2Interface, "");
-		m_context.setSolver(smtlib2Interface->smtlib2Interface());
-	}
-	m_context.clear();
-	m_context.setAssertionAccumulation(false);
-
 	resetSourceAnalysis();
 
-	m_genesisPredicate = createSymbolicBlock(arity0FunctionSort(), "genesis");
-	addRule(genesis(), "genesis");
-
 	set<SourceUnit const*, IdCompare> sources;
 	sources.insert(&_source);
-	for (auto const& source: _source.referencedSourceUnits(true))
+	for (auto const& source: _source.referencedSourceUnits())
```

Comment:
It should not add sources recursively because later they are visited via third loop?

Reply:
Hm maybe they should be added recursively though

Reply:
Can this be tested somehow?

Reply:
the added tests
`test/libsolidity/smtCheckerTests/imports/imported_fail_1.sol`
and
`test/libsolidity/smtCheckerTests/imports/imported_fail_2.sol`
have 3 sources where one imports the next, and this worked fine. Not sure which case would brake it. I changed it back to `recursive = true` to be safe.

Reply:
I am fine with leaving true